File: solutions/cpp/0077-combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> result;
        vector<int> combination(k);
        generateCombinations(1, n, k, combination, result);
        return result;
    }

private:
    void generateCombinations(int start, int n, int k, vector<int> &combination, vector<vector<int>> &result) {
        if (k == 0) {
            result.push_back(combination);
            return;
        }
        for (int i = start; i <= n; ++i) {
            combination[combination.size() - k] = i;
            generateCombinations(i + 1, n, k - 1, combination, result);
        }
    }
};

File: solutions/csharp/0077-combinations.cs
public class Solution {
    public IList<IList<int>> Combine(int n, int k) {
        var result = new List<IList<int>>();
        GenerateCombinations(1, n, k, new List<int>(), result);
        return result;
    }
    
    private void GenerateCombinations(int start, int n, int k, List<int> combination, IList<IList<int>> result) {
        if (k == 0) {
            result.Add(new List<int>(combination));
            return;
        }
        for (var i = start; i <= n; ++i) {
            combination.Add(i);
            GenerateCombinations(i + 1, n, k - 1, combination, result);
            combination.RemoveAt(combination.Count - 1);
        }
    }
}
